Apply sky brightness multiplier in correct color space

Kaleb-Reid opened the pull request:
In the RD renderers, if we are using fog and a solid colour background, ```clear_color``` should not be in linear encoding before being passed to the shader, as this will be converted again later. If we are using a solid colour background and not using fog, the energy needs to be multiplied _after_ converting to linear encoding. In scenarios where the sky is drawn and we are not using fog or not using solid colour background, the multiplication happens in the shader.

Compatibility is similar except after the energy is multiplied in, ```clear_color``` needs to be converted back to non-linear srgb encoding. This pr also moves the energy multiplication in the shader to be after converting the fragment colour to linear encoding.

This should:
- Apply brightness multipliers for the sky on colours with the correct color encoding.
- Fix applying the brightness multiplier twice with solid colour background and fog.
- Fix incorrectly passing ```clear_color``` with linear encoding to sky material in the RD renderers, which converts it again later.

Fixes https://github.com/godotengine/godot/issues/108779

Kaleb-Reid committed:
Apply sky brightness multiplier in correct color space

Kaleb-Reid commented:
@allenwp I updated the pr with a better description of what is happening. Where I removed the ```srgb_to_linear``` call, ```clear_color``` is in fact encoded using the nonlinear sRGB transfer function. The issue is that converting this now is incorrect because it will be converted a second time later when it is put into the uniform buffer for the sky shader.

allenwp commented:
> @allenwp I updated the pr with a better description of what is happening. Where I removed the ```srgb_to_linear``` call, ```clear_color``` is in fact encoded using the nonlinear sRGB transfer function. The issue is that converting this to linear now is incorrect because it will be converted a second time later when it is put into the uniform buffer for the sky shader.
> 
> 
> 
> Not sure if the comments should still be added given that the value is not being passed in linear encoding.

Ah, so there is another problem that colour/lighting maths are being performed on these nonlinear sRGB-encoded values (when they are multiplied by the energy multiplier). This is likely incorrect and should be addressed.

To put it simply, no maths of any sort should be performed on nonlinear sRGB-encoded values. The only exception is if you are using the math as a layer of "encoding" that will be directly reversed before doing any further operations, like with a "luminance multiplier".

Kaleb-Reid commented:
@allenwp I've updated the pr to correct this. If fog is enabled, ```bg_energy_multiplier``` is already applied to linear colours in the shader via ```sky_energy_multiplier```, so it doesn't need to be multiplied into the colour before the shader. For the cases where the sky is not drawn (clear colour / background colour with no fog), the colour is converted to linear encoding, multiplied, and converted back.

allenwp commented:
Thanks! It will take me a while to review this, and I'm not familiar with this part of the code base, but at a glance, I believe the intent of this updated PR is correct.

A couple of things that I can mention before doing a thorough review: dithering must always be applied as the final step before quantization. The dithering in the sky shaders is implemented in quite a weird way, and probably can be improved, but for now it should be left as-is, as the final step in the shader.

Another thing to mention is that conversions to and from nonlinear sRGB encoding are expensive and should be minimized. So if there's a way to keep things in linear (and document it with comments) the whole way through to limit conversions back and forth, that is preferred for performance reasons.

Finally, the Compatibility renderer intentionally performs some colour maths incorrectly for performance reasons, so avoiding decoding and re-encoding in this renderer may actually be the intended design, even though it produces incorrect results. You can see [this comment](https://github.com/godotengine/godot/issues/90259#issuecomment-2704992993) for more details about this. Conversely, this sort of optimization to perform colour maths on nonlinear sRGB-encoded values should never happen in the Mobile/Forward+ renderers.

Anyway, at some point I'll take a deeper look into this PR to verify it.

And thanks for prompting me to write all of this out. Our discussion here is a good template for some colour documentation for contributors...

Kaleb-Reid commented:
@allenwp I can probably optimize the RD renderers and remove the extra conversions, as after what I added they are only accessed in linear anyways.

As for Compatibility, I wasn't originally going to add the conversions but I saw that other environment colours (e.g. ambient light) are converted to linear before energy calculations, and it's also done on the cpu and not the gpu so I decided to add it anyways. In terms of the gpu work, I only rearranged the order of calculations to multiply with the energy while the colour is in linear and did not add anything. I did move the dithering before the conversion back to srgb because it uses the energy multiplier and I wasn't sure if it mattered.

If parts / all of Compatibility changes need to be removed / reworked I can do that. I will start with moving the dithering back to where it was.

allenwp commented:
>and it's also done on the cpu and not the gpu so I decided to add it anyways.

Great; this sounds entirely reasonable and does not go against the spirit of the Compatibility renderer!

allenwp commented:
> If parts / all of Compatibility changes need to be removed / reworked I can do that. I will start with moving the dithering back to where it was.

Based on what you said, the dithering change is the only part that should be reverted. All the rest sounds good and doesn't need to be changed!

In the future, we can improve the dithering to not depend on the energy multiplier or revisit this to come to some other sort of solution...

allenwp commented on drivers/gles3/shaders/sky.glsl:
```diff
@@ -264,10 +264,10 @@ void main() {
-	frag_color.a = alpha;
-
 #ifdef USE_DEBANDING
-	frag_color.rgb += interleaved_gradient_noise(gl_FragCoord.xy) * sky_energy_multiplier * luminance_multiplier;
+	color += interleaved_gradient_noise(gl_FragCoord.xy) * sky_energy_multiplier;
 #endif
+
+	frag_color.rgb = color * luminance_multiplier;
+	frag_color.a = alpha;
```
Ah, I see! I hadn't realized this was mathematically identical. This is why it looked like you were replacing the existing code with something that was equivalently incorrect :)

In terms of code readability and git logs, it's probably not as good to have this change, even though it's higher performance: I think this sort of optimization and correcting/improving debanding is better suited for a different PR.

allenwp reviewed (approved):
Code looks correct and I have confirmed this PR fixes #108779 for Forward+, Mobile, and Compatibility (with a background energy multiplier) for HDR 2D enabled and disabled.

Thanks!

Kaleb-Reid commented:
I'm not sure what difference it makes but I believe the energy multiplier should always be multiplied into the clear colour after sending it to the shader so that the framebuffer is cleared with the correct color so I've changed that.

Calinou reviewed (approved):
Tested locally (rebased on top of `master` https://github.com/godotengine/godot/commit/833889aec024b792b1c5ce5089eaebc5eb45692f), it works as expected. Code looks good to me.

### Before

https://github.com/user-attachments/assets/066900dd-292a-43b1-90b6-df0c12169674

### After

https://github.com/user-attachments/assets/ad0527e7-b2cf-4cf6-9f84-9e768a787745

blueskythlikesclouds reviewed (approved):
Looks good, it all seems consistent now.

Repiteo commented:
Needs rebase

Repiteo commented:
Thanks!
